### PR TITLE
[FrameworkBundle][Serializer] Allow serializer default context configuration

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no 
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | todo if we go with it 

This allow to configure the `context` for normalizer's on a high level:

```yaml
serializer:
    default_context:
        enable_max_depth: true
```

This feature was asked in https://github.com/api-platform/core/issues/1764#issuecomment-431272313 and I thought it could be in symfony instead. The messenger component has the same configuration.
This re-opens https://github.com/symfony/symfony/pull/28927 (for discussion). I'm still unsure about the discussion in https://github.com/symfony/symfony/pull/28927#discussion_r258092928 and what solution I should implement. Please let me know if you come to an agreement.

### CODE REVIEW

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php [hunk 1]
```diff
@@ -50,11 +50,39 @@ public function process(ContainerBuilder $container)
...
+
+            $reflection = new \ReflectionClass($class);
+
+            if (null === $constructor = $reflection->getConstructor()) {
+                continue;
+            }
+
+            foreach ($constructor->getParameters() as $arg) {
+                if ('defaultContext' === $arg->name) {
+                    $definition->setArgument('$defaultContext', '%serializer.default_context%');
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: Maybe only if it's not already set (to allow the user overwrite this)?

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php (hunk 1 above)
Comment (soyuka) [reply src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: I may use `bind->('$defaultContext', '...');` in the serializer maybe?

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php [hunk 2]
```diff
@@ -50,11 +50,39 @@ public function process(ContainerBuilder $container)
...
+            $definition = $container->getDefinition($service);
+            if (!$definition->isAutowired()) {
+                continue;
+            }
+
+            if (null === $class = $definition->getClass()) {
+                continue;
+            }
+
+            $reflection = new \ReflectionClass($class);
```
Comment (chalasr) [inline src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]:
```suggestion
            $reflection = $container->getReflectionClass($class);
```

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php [hunk 3]
```diff
@@ -50,11 +50,39 @@ public function process(ContainerBuilder $container)
...
 
         if (!$encoders = $this->findAndSortTaggedServices($this->encoderTag, $container)) {
             throw new RuntimeException(sprintf('You must tag at least one service as "%s" to use the "%s" service.', $this->encoderTag, $this->serializerService));
         }
 
         $serializerDefinition->replaceArgument(1, $encoders);
+        $this->addDefaultContextParameter($encoders, $container);
+    }
+
+    private function addDefaultContextParameter($services, $container)
```
Comment (chalasr) [inline src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: type declarations are missing

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: Using bindings sounds better, `$definition->setBindings(['$defaultContext', new BoundArgument($defaultContext, false)] + $value->getBindings());`.

Also I'm not fond of exposing configuration values as container parameters, those are not extension points. The parameter should be resolved before injection and be removed from the container once we are done with it.

**Review by chalasr — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
@soyuka what is the state of this PR ?

**Comment by soyuka** [comment]
@stof I'm wondering as well, I'm really not sure that we've come to an agreement on what needs to be done. If the community knows exactly how we can go on, it'll be my pleasure to continue working on it! 

Do we actually want this feature?

**Comment by norkunas** [comment]
I'd like to use this feature. Currently to always expose extended datime format I override datetime normalizer context in compiler pass.

**Comment by dunglas** [comment]
I'm in favor of merging this feature

**Comment by lyrixx** [comment]
Me too. @soyuka do you have time to finish it, or do you need help?

**Comment by soyuka** [comment]
No problem I'll finish this. About the code I should use `setBinding` then? Won't that be a possible conflict with userland? If someone declares a $defaultContext it'll overwrite ours right?

**Comment by lyrixx** [comment]
> No problem I'll finish this. About the code I should use `setBinding` then? Won't that be a possible conflict with userland? If someone declares a $defaultContext it'll overwrite ours right?

yes 👍🏼

**Review by lyrixx — APPROVED** [review]
LGTM 👍🏼

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @soyuka.

**Pull request merged**